### ISSUE TITLE
torch.fft.irfft(n) doesn’t handle non-Hermitian inputs in a consistent way

### ISSUE THREAD

**Issue by pavel-shmakov** [pytorch/pytorch#154496]
### 🐛 Describe the bug

Example:

```python
n_points = 2**13
device = "cuda"
y = torch.ones(n_points, dtype=torch.float32, device=device)
f = torch.fft.rfftn(y, norm="forward") # Same for rfft/irfft
f[0] += 100j
g = torch.fft.irfftn(f, norm="forward", s=y.shape)
print(g[:10])
```

On CPU we get `tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])`, consistent with the [[documentation](https://docs.pytorch.org/docs/stable/generated/torch.fft.irfft.html#torch.fft.irfft)](https://docs.pytorch.org/docs/stable/generated/torch.fft.irfft.html#torch.fft.irfft) saying that the imaginary part is ignored for frequencies that shouldn’t have them (i.e. zeroth mode).

However, on GPU we get `tensor([-99., 101., -99., 101., -99., 101., -99., 101., -99., 101.])` . We can reproduce this only in 1D, and only for `2**13` or `2**14` points. It does reproduce for different `norm`s, though one has to scale the added imaginary part accordingly.

It would be also good to have the behaviour for non-Hermitian symmetric inputs documented in general. For example, we found that in 2D there seems to be symmetrisation happening:

```python
n_points = 8192
device = "cuda"

y = torch.zeros((n_points, n_points)).to(device)
f1 = torch.fft.rfftn(y)
f1[1, 0] -= 5j
f1[-1, 0] += 10j
g1 = torch.fft.irfftn(f1, s=y.shape)

f2 = torch.fft.rfftn(y)
f2[1, 0] -= 7.5j
f2[-1, 0] += 7.5j
g2 = torch.fft.irfftn(f2, s=y.shape)
torch.testing.assert_close(g1, g2)  # Passes
```

However, it’s not clear whether we can rely on this behaviour.

**Why is this important**: in some of the Neural Operator architectures, it takes special care to achieve Hermitian symmetry. Consider [[FNO](https://arxiv.org/pdf/2010.08895)](https://arxiv.org/pdf/2010.08895) and its implementation in [[neuraloperator](https://github.com/neuraloperator/neuraloperator)](https://github.com/neuraloperator/neuraloperator) library: the signal gets transformed to the Fourier domain, then multiplied by a [[complex-valued learned weight tensor](https://github.com/neuraloperator/neuraloperator/blob/main/neuralop/layers/spectral_convolution.py#L341)](https://github.com/neuraloperator/neuraloperator/blob/main/neuralop/layers/spectral_convolution.py#L341), and finally transformed back. That multiplication results in the loss of Hermitian property. Fixing it on Python level may introduce noticeable overhead though.

### Versions

PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.10 | packaged by conda-forge | (main, Apr 10 2025, 22:21:13) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.10.236-228.935.amzn2.x86_64-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 550.163.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             7
BogoMIPS:                             4999.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke avx512_vnni
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            128 KiB (4 instances)
L1i cache:                            128 KiB (4 instances)
L2 cache:                             4 MiB (4 instances)
L3 cache:                             35.8 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] gpytorch==1.14
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch_finufft==0.1.0
[pip3] tensorly-torch==0.5.0
[pip3] torch==2.7.0
[pip3] torch-geometric==2.6.1
[pip3] torch_scatter==2.1.3.dev1+pt27cu128
[pip3] torchmetrics==1.7.1
[pip3] triton==3.3.0
[conda] gpytorch                  1.14                     pypi_0    pypi
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.2                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
[conda] pytorch-finufft           0.1.0                    pypi_0    pypi
[conda] tensorly-torch            0.5.0                    pypi_0    pypi
[conda] torch                     2.7.0                    pypi_0    pypi
[conda] torch-geometric           2.6.1                    pypi_0    pypi
[conda] torch-scatter             2.1.3.dev1+pt27cu128          pypi_0    pypi
[conda] torchmetrics              1.7.1                    pypi_0    pypi
[conda] triton                    3.3.0                    pypi_0    pypi
